openshift/assisted-service
# MGMT-13890: Drain before deprovisioning spoke nodes

This PR drains the spoke node before the BMH is configured for deprovisioning.

A few annotations are added to control and monitor this process:

- `bmac.agent-install.openshift.io/drain-started-at`: Set by BMAC when the drain process starts - this is used by BMAC to determine when the drain process has timed out.
- `bmac.agent-install.openshift.io/drain-timeout`: Set by the user in `time.Duration` format (eg `5m`). Used by BMAC to determine when the drain process has timed out.
- `bmac.agent-install.openshift.io/drain-status`: Set by BMAC to inform the user of the status of the drain process. Will be one of the following constant strings, or the error string returned from the drain function

```
drainStatusSuccess    = "drain succeeded"
drainStatusInProgress = "draining in progress"
drainStatusTimeout    = "drain timed out"
```

The `Drainer` interface is created to allow for easier unit testing, but it otherwise just a passthrough to the `kubectl` functions.

## List all the issues related to this PR

https://issues.redhat.com/browse/MGMT-13890


## What environments does this code impact?


## How was this code tested?

Manually in dev-scripts.

- Created a 5 node spoke cluster
- Set `bmac.agent-install.openshift.io/remove-agent-and-node-on-delete` on the BMH
- Delete the BMH

Observe the node drain before deleting

Also tested with the timeouts by attempting to delete the other worker node which would result in drain timing out as the ingress pods had no way to be rescheduled.




## Checklist

- [ ] This change does not require a documentation update (docstring, `docs`, README, etc) TODO

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?
